src/ProductConstructionService/ProductConstructionService.Api/Queue/PcsJobsProcessor.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+
+using System.Text.Json;
+using System.Threading;
+using Azure.Storage.Queues;
+using Azure.Storage.Queues.Models;
+using ProductConstructionService.Api.Queue.WorkItems;
+
+namespace ProductConstructionService.Api.Queue;
+
+public class PcsJobsProcessor(
+    ILogger<PcsJobsProcessor> logger,
+    PcsJobProcessorOptions options,
+    PcsJobsProcessorStatus status,
+    QueueServiceClient queueServiceClient)
+    : BackgroundService
+{
+    private readonly ILogger<PcsJobsProcessor> _logger = logger;
+    private readonly PcsJobProcessorOptions _options = options;
+    private readonly PcsJobsProcessorStatus _status = status;
+
+    private const int JobInvisibilityTimeoutSeconds = 30;
+
+    protected override async Task ExecuteAsync(CancellationToken cancellationToken)
+    {
+        await Task.Run(async () => {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                await _status.Semaphore.WaitAsync(cancellationToken);
+                if (!cancellationToken.IsCancellationRequested)
+                {
+                    await ProcessJobs(cancellationToken);
+                }
+            }
+        }, cancellationToken);
+    }
+
+    private async Task ProcessJobs(CancellationToken cancellationToken)
+    {
+        QueueClient queueClient = queueServiceClient.GetQueueClient(_options.QueueName);
+        _logger.LogInformation("Starting to process PCS jobs {queueName}", _options.QueueName);
+        while (!cancellationToken.IsCancellationRequested && _status.ContinueWorking)
+        {
+            try
+            {
+                var pcsJob = await GetPcsJob(queueClient, cancellationToken);
+
+                if (pcsJob == null)
+                {
+                    // Queue is empty, wait a bit
+                    _logger.LogInformation("Queue {queueName} is empty. Sleeping for {sleepingTime} seconds", _options.QueueName, _options.EmptyQueueWaitTime);
+                    await Task.Delay(_options.EmptyQueueWaitTime, cancellationToken);
+                    continue;
+                }
+
+                ProcessPcsJob(pcsJob);
```

Comment:
Should we not set here the status to "working"?

Reply:
perhaps `ContinueWorking` should just be `Work`

Reply:
How will we know if we have finished the last item already?

Reply:
Last item in the queue? If the queue is empty, the `JobProcessor` will go into `Task.Delay`, and then keep checking for queue items periodically.
If we issue a `Stop` command during this period, the `JobProcessor` will finish waiting for the delay, and then exit the loop